biocrnpyler/dna_part_misc.py

```diff
@@ -109,52 +110,66 @@ def update_component(self,internal_species=None,**keywords):
+                del othersite.linked_sites[self_tuple]
+                othersite.linked_sites[(newcomp,intermolecular)] = mystuff
             return newcomp
         else:
-            for othersite in self.linked_sites:
+            for othersite_tpl in self.linked_sites:
+                othersite = othersite_tpl[0]
+                intermolecular = othersite_tpl[1]
                 populate = True #by default, add the appropriate data members to the other site
-                if(self.count_inputs(othersite)==1):
+                if(intermolecular==0):
                     #the reaction with the site in question is intramolecular
                     #that means we should only populate the other site if our internal_species
                     #has the proper location bound by integrase
                     #if the reaction is intramolecular, that means that this component knows everything in order to decide
                     #create the reaction.
                     #the linked site is populated because then the linked site knows not to create the reaction.
                     #after all, there is one reaction per two sites
-                    otherisbound = othersite.integrase in internal_species.parent[othersite.position]
+                    
+                    if(self.integrase_binding):
+                        otherisbound = othersite.integrase in internal_species.parent[othersite.position]
+                    else:
+                        otherisbound = not isinstance(internal_species.parent[othersite.position],ComplexSpecies)
                     if(not otherisbound):
                         populate = False
                 if(populate):
                     #if the reaction is intermolecular then the linked site is populated.
                     #this means only a fully populated site would have all the information to create the reaction
                     #once again this results in one out of two sites that actually outputs a "reaction" object, as required
-                    mystuff = copy.copy(othersite.linked_sites[self])
-                    del othersite.linked_sites[self]
-                    othersite.linked_sites[newcomp]=mystuff
-                    assert(othersite in newcomp.linked_sites)
+                    mystuff = copy.copy(othersite.linked_sites[(self,intermolecular)])
+                    del othersite.linked_sites[(self,intermolecular)]
+                    othersite.linked_sites[(newcomp,intermolecular)]=mystuff
+                    assert((othersite,intermolecular) in newcomp.linked_sites)
             return newcomp
-    def count_inputs(self,site):
-        assert(site in self.linked_sites)
-        dict_inputs = []
-        for transformation in self.linked_sites[site][0]:
-            for pspec in transformation.parentsdict:
-                dict_inputs += [transformation.parentsdict[pspec]]
-        return len(set(dict_inputs))
+    def update_species(self):
```

Comment:
part of the problem is the way that i am indicating that an integrase reaction can happen, and what the products are. A site checks if the other site has integrase bound. If it does, then the reaction can proceed. This is because I want a construct with multiple integrase sites to properly allow the sites to be bound and then a recombination to occur all without disturbing integrase bound to non-reacting sites. So that means the idea of integrase binding to sites is kind of separate from the idea of integrase flipping the sites. 

I suppose I could make it so that the presence of a binding mechanism in the site determines whether it binds the integrase or not. So then it would have a binding mechanism and a recombination mechanism. But normally if you don't have a mechanism that means you inherit it from the mixture, right? So I think doing it this way would go against the way it's done in other components.

Also the way integrase_function s work now is by replacing attP with attL (for example) and not really caring about the status of that attP site (was it bound by integrase before? If so, it should be bound when replaced). So then you need this keyword to know whether the new site should be bound or not. This does seem fixable.

Reply:
I think it would be great if all the binding could happen in the mechanism. It seems like if the Mechanism knows the previous conformation, the sites that are involved, and the final conformation, this should be sufficient for deciding what reactions get from input-->output.